clippy_lints/src/casts/dangling_ptr.rs

```diff
@@ -0,0 +1,41 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::SpanRangeExt;
+use clippy_utils::{expr_or_init, is_expr_positive_literal, std_or_core};
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, Mutability, Ty, TyKind};
+use rustc_lint::LateContext;
+
+use super::DANGLING_PTR;
+
+pub(super) fn check(cx: &LateContext<'_>, expr: &Expr<'_>, from: &Expr<'_>, to: &Ty<'_>) {
+    if let TyKind::Ptr(ref mut_ty) = to.kind {
+        let init_expr = expr_or_init(cx, from);
+        if is_expr_positive_literal(init_expr)
+            && let Some(std_or_core) = std_or_core(cx)
+            && let Some(var) = from.span.get_source_text(cx)
+        {
+            let (msg, sugg_fn) = match mut_ty.mutbl {
+                Mutability::Not => (format!("`{var} as *const _` detected"), "ptr::dangling"),
+                Mutability::Mut => (format!("`{var} as *mut _` detected"), "ptr::dangling_mut"),
```

Comment:
Please don't copy the expression into the message. It's already highlighted with the span. Just a simple "manual creation of a dangling pointer" is enough.